src/database/helpers/openDB.ts

```diff
@@ -18,6 +18,46 @@ export function createDBAccess<DBSchema>(opener: () => Promise<IDBPDatabase<DBSc
         return db
     }
 }
+export function createDBAccessWithAsyncUpgrade<DBSchema, AsyncUpgradePreparedData>(
+    firstVersionThatRequiresAsyncUpgrade: number,
+    latestVersion: number,
+    opener: (currentTryOpenVersion: number, knowledge?: AsyncUpgradePreparedData) => Promise<IDBPDatabase<DBSchema>>,
+    asyncUpgradePrepare: (db: IDBPDatabase<DBSchema>) => Promise<AsyncUpgradePreparedData | undefined>,
+) {
+    let db: IDBPDatabase<DBSchema> | undefined = undefined
+    let pendingOpen: Promise<IDBPDatabase<DBSchema>> | undefined = undefined
+    async function open(): Promise<IDBPDatabase<DBSchema>> {
+        OnlyRunInContext(['background', 'debugging'], 'Database')
+        if (db?.version === latestVersion) return db
+        let currentVersion = firstVersionThatRequiresAsyncUpgrade
+        let lastVersionData: AsyncUpgradePreparedData | undefined = undefined
+        while (currentVersion < latestVersion) {
+            try {
+                db = await opener(currentVersion, lastVersionData)
+                // if the open success, the stored version is small or eq than currentTryOpenVersion
+                // let's call the prepare function to do all the async jobs
+                lastVersionData = await asyncUpgradePrepare(db)
+            } catch (e) {
+                if (currentVersion >= latestVersion) throw e
+                // if the stored database version is bigger than the currentTryOpenVersion
+                // It will fail and we just move to next version
+            }
+            currentVersion += 1
+            db?.close()
+            db = undefined
+        }
+        db = await opener(currentVersion, lastVersionData)
+        db.addEventListener('close', (e) => (db = undefined))
+        if (!db) throw new Error('Invalid state')
+        return db
+    }
+    return () => {
+        // Share a Promise to prevent async upgrade for multiple times
+        if (pendingOpen) return pendingOpen
```

Comment:
Attention: If an error was thrown in the `open`, the rejected promise object will always be returned.

Reply:
Nice catch